components/core/src/fs.rs

```diff
@@ -719,6 +749,11 @@ impl AtomicWriter {
     /// temporary file to ensure all data is flushed to disk and then
     /// renaming the file into place.
     fn finish(self) -> io::Result<()> {
+        if let Some(permission) = self.permission {
+            set_permissions(&self.tempfile.path(), permission).map_err(|e| {
+                io::Error::new(io::ErrorKind::Other, e.to_string())
+            })?;
```

Comment:
I don't love this.  We made a decision when creating the API to return an `io::Result<()>` here to make it easier on callers.  I don't have enough Rust experience to know if there is a better way to achieve the same ease of use so I'm happy to take whatever alternative y'all have (or just leave this bit of grossness).

Reply:
You could implement `From<Error>` for `io::Error`

```rust
impl From<Error> for io::Error {
    fn from(e: Error) -> Self {
        io::Error::new(io::ErrorKind::Other, e.to_string())
    }
}
```

and then you would just need 
```rust
if let Some(permission) = self.permission {
    set_permissions(&self.tempfile.path(), permission)?;
}
```

You could even match on the `Error` type and try and convert it to the corresponding `io::Error` (`Error::PermissionFailed` to `io::Error::PermissionDenied`). But that just leads to the question, why does `set_permission` not use `io::Result`?

With all that said, the "correct" answer probably involves refactoring outside the scope of this PR so I am totally fine leaving it as is.

Reply:
I think the "correct" answer lies in https://github.com/habitat-sh/habitat/issues/6557, but that is most assuredly outside the scope of this PR :smile: 

I'm OK with leaving this as-is for the time being.